README.md
# deno-api-server

An http/rest api server for deno. Based on std deno http library and use the
concept of functional programming for your endpoint definitions.

- [API Documentation](https://doc.deno.land/https/deno.land/x/deno_api_server/mod.ts)
- [Preset API Documentation](https://doc.deno.land/https/deno.land/x/deno_api_server/src/presets/mod.ts)
- [Testing API Documentation](https://doc.deno.land/https/deno.land/x/deno_api_server/dev_mod.ts)
- [Release notes](./RELEASE_NOTES.md)

## Basic functionality example

```typescript
import {
  Api,
  EMethod,
  Route,
} from "https://deno.land/x/deno_api_server/mod.ts";

const api = new Api({ port: 8080 });

api.addRoute(
  new Route(EMethod.GET, "/")
    .addPipe(({ response }) => {
      response.body = { message: "Hello API" };
    }),
);

console.log(`Start server localhost:${api.serverConfig.port}`);
await api.listen();
```

## events

`API_ADD_ROUTE` (RouteEvent) will trigger before a new route append to api
stack. Tip use this as injection hook for your services

`BEFORE_ROUTE` (RouteEvent) will trigger before execute the current route

`BEFORE_REQUEST` (RequestEvent) will trigger before a route are detected and
before it execute

`ROUTE_NOT_FOUND` (RequestEvent) will trigger if route not match [404]

`ROUTE_ERROR` Error on route execute [500]
`CRITICAL_ERROR` Critical error on response execution [500]

## More code examples

### Add simple route to your api

```typescript
/** INFO: don't forget route to your Api, see first example **/
new Route(EMethod.GET, '/say-hello')
    .addPipe(({response}) => {
        response.body = {message: 'Hello API'};
    })
)
```

### Add simple route with multi methods

```typescript
/** INFO: don't forget route to your Api, see first example **/
new Route([EMethod.GET, EMethod.POST], "/say-hello")
  .addPipe(({ response, request }) => {
    response.body = { message: `Hello API by ${request.method}` };
  });
```

### Use state to pass data between pipes

```typescript
/** INFO: don't forget route to your Api, see first example **/
new Route(EMethod.GET, "/state")
  .addPipe(({ response, state }) => {
    const before = new Date();
    response.body = { before };
    state.set("called", before);
  })
  .addPipe(({ response, state }) => {
    Object.assign(response.body, {
      passed: state.get("called"),
      done: new Date(),
    });
  });
```

### Throw error to handle by api

```typescript
/** INFO: don't forget route to your Api, see first example **/
new Route(EMethod.GET, "/error")
  .addPipe(() => {
    throw new RequestError("I dont like", 400);
  });
```

### Key mapped uri with KeyMatch for url params

```typescript
new Route(
  EMethod.GET,
  new KeyMatch(
    "/get-by-key-name/:id/:name",
    {
      id: { type: "number" },
      name: {},
    },
  ),
)
  .addPipe(({ response, match }) => {
    const { params } = match;
    response.body = {
      message: `You call with keymatch`,
      id: params.get("id"),
      name: params.get("name"),
    };
  });
```

For more details about KeyMatch see API Documentation KeyMatch

# API Dokumentation

### class Api

Core server class, create an instance for your own api

Similar configuration of deno http service

```
new Api(config: IServerConfig)
```

Add new Routes / Entpoints

```
api.addRoute(route: IRoute)
```

Start server to listen

```
api.listen();
```

### class Route

Core route or entpoint defintion for your api.

```
new Route(method: string[] | string, uri : IMatcher | string)
```

Add route pipeline

```
route.addPipe(pipe: IPipe) : Route
```

Example: Add a stack of pipes

```
route
    .addPipe((context: IContext) => { ... })
    .addPipe(async (context: IContext) => { ... })
```

`Tip` use deconstruction for select context props

Execute a route pipeline, this will be called by api if route match with uri
defintion

```
route.execute(url: URL, request: ServerRequest, response: IResponse): Promise<IContext>
```

## Matcher

### class UriMatch

Base endpoint matcher for simple definition of uris.

`INFO` you can use a string for Route definition, the Route will create a simple
UriMatch

## class KeyMatch

More complex uri definition for your route with transpile your parameters. A
KeyMatch can use a transform function to change the type of value

```
new KeyMatch(uri: string, describe: IKeyDescribes)
```

Examples:

```
new KeyMatch('/get/:id', { id: {} }) => match: /get/1, params: { id: '1' }
new KeyMatch('/get/:id', { id: {} }) => match: /get/hello-my, params: { id: 'hello-my' }

// with transpile type
new KeyMatch('/get/:id', { id: { type: 'number' } }) => match: /get/1, params: { id: 1 }
new KeyMatch('/get/:id', { id: { type: 'number' } }) => not match: /get/hello

// with rest type
new KeyMatch('/get/:path', { path: { type: 'rest' } }) => match /get/super-dup/megaman, params: { path: 'super-dup/megaman' }
```

### IKeyDescribe.type

A list of `EPatternTypes`

- `Any` for any parameter like string, number and combination
- `Number` for float and numbers
- `Int` for integer numbers
- `Alpha` for alphabetic values
- `Hash` for hash values
- `Rest` for rest of the url path

### Custom methode to define a custom param pattern

Use Property `describe` to define a `IPatternDescribe`

Example:

```js
new KeyMatch('/get/:p', { 
  p: {
    describe: {
      pattern: '(a-z){1}',
      transform: (v: string) => v.toUpperCase()
    }
  }
}) => will match /get/g, params: { p: 'G' }
```

## Quick test api server

```typescript
import "https://deno.land/x/deno_api_server@v0.0.2/example.ts";
/**
 * will start server on localhost:8080
 * try all entpoints on example
 **/
```

All presets defined in `src/presets`

## List of presets

[Preset API Documentation](https://doc.deno.land/https/deno.land/x/deno_api_server/src/presets/mod.ts)

All presets defined in `src/presets`


### Pipes presets `src/presets/pipes/*`

#### `body/raw-body.pipe.ts`

Read body data of request.

Set state `body` with string data Set state `bodyType` with the type of data ==
raw

```typescript
import rawBodyPipe from "https://deno.land/x/deno_api_server/src/presets/pipes/body/raw-body.pipe.ts";
route
  .addPipe(rawBodyPipe)
  .addPipe(({ state }: IContext) => {
    const anyBody = state.get("body");
  });
```

#### `body/json-body.pipe.ts`

Try to parse json data of request body. Otherwise throw 400 error.

Set state `body` with json data Set state `bodyType` with the type of data ==
json

```typescript
import jsonBodyPipe from "https://deno.land/x/deno_api_server/src/presets/pipes/body/json-body.pipe.ts";
route
  .addPipe(jsonBodyPipe)
  .addPipe(({ state }: IContext) => {
    const validJsonBody = state.get("body");
  });
```

#### `process/redirect.pipe.ts`

Redirect pipe with default 302 redirection to your url. Basic implementation to
redirect your route.

```typescript
import redirectPipe from "https://deno.land/x/deno_api_server/src/presets/pipes/process/redirect.pipe.ts";
route
  .addPipe(redirectPipe("/other-page"));
```

If you want to redirect a dynamic page use any capsule pipe function, like here

```typescript
import redirectPipe from "https://deno.land/x/deno_api_server/src/presets/pipes/process/redirect.pipe.ts";
route
  .addPipe((context: IContext) => {
    const url = `/test-to-my?set=1`;
    return redirectPipe(url)(context);
  });
```

## Testing

Use buildin mocks to create a api request.

[Build-In](https://deno.land/x/deno_api_server/dev_mod.ts)
[API Documentation](https://doc.deno.land/https/deno.land/x/deno_api_server/dev_mod.ts)

A simple example

```typescript
import { mockApi } from "https://deno.land/x/deno_api_server/dev_mod.ts";

Deno.test("Example mockApi route success request", async () => {
  const route = new Route("GET", "/hello");

  // create api
  const api = mockApi(route);

  await api.sendByArguments("GET", "/hello");

  assertEquals(api.lastRoute === route, true);
  assertEquals(api?.lastContext?.response.status, 200);
});
```

A payload example

```typescript
import {
  mockApi,
  mockRequest,
} from "https://deno.land/x/deno_api_server/dev_mod.ts";

Deno.test("Example mockApi post request with request data", async () => {
  const route = new Route("POST", "/submit");
  route
    .addPipe(jsonBodyPipe)
    .addPipe(({ state, response }) => {
      response.status = 201;
      response.body = state.get("body");
    });

  const api = mockApi(route);

  const request = mockRequest("POST", "/submit", {
    name: "super",
  });
  await api.sendByRequest(request);

  assertEquals(api.lastRoute === route, true);
  assertEquals(api?.lastContext?.response.status, 201);
  assertEquals(api?.lastContext?.response.body, { name: "super" });
});
```

More examples in `example/unit-testing.test.ts`

## Best Practice

Tips and best practice to use this server.

- Write custom and small pipes to reuse
- create a custom createRoute function to setup your app or use events to inject
  services
- use Route.di / Route.injections to share services and other stuff
- create factory functions for injections, like getConnection factory to connect
  to database only if route will execute
- write test by mock your injections (like mock service for your database
  connection)
- use state to pass data between your pipes

## examples

See `example` folder for more use case examples.

#### `example/unit-testing.test.ts`

route and api testing example and how to use build-in mock functions.

### Third party examples are

#### example/body-validation.ts

Use validasaur to validate your json body
[validasaur deno.land](https://deno.land/x/validasaur)

#### example/authentification-jwt.ts

Use djwt to handle and secure your endpoints, a very simple example to customize
your process [validasaur deno.land](https://deno.land/x/djwt)

## links

[deno](https://deno.land)

## Plugins
`info` import the plugin to your api main and regist by call the default function.

```
import plugin from 'PLUGIN.ts'

const api = new Api();
plugin(api, { /* plugin configuration */ })
```

### Swagger json
use plugins/swagger, but still in WIP

### healthcheck
use plugins/healthcheck add add this plugin like

`depricated since 0.6.0` preset/health

### status
use plugins/status to show more details about server status, you can add more details for your endpoint by use body oder the custom handler.

`depricated since 0.6.0` preset/status

### access-log `since 0.6.0`
use plugins/status to show all request actions, optional filtering of checks


### add-route `since 0.6.1`
use plugins/add-route to show all route registration by using a custom log handler






RELEASE_NOTES.md
# Release notes

## 0.6.2
add more options for plugin access log

## 0.6.1
add route plugin
fix some typescipt issues


## 0.6.0
add access logger method

`assertCalledCount` new testing function for mockFn
`assertCalledWith` new testing function for mockFn
`assertCalledWithAt` new testing function for mockFn

### Depricated
`preset/healthz` use plugin/healthcheck
`preset/status` use plugin/status

## 0.5.0
`formBodyPipe` pipe to parse form data

### Depricated
Obsolate authentification example

## 0.4.0
Upgrade to deno version 1.25.2
Update all std to 0.115.0

`rawBodyPipe` fill state with body string
`jsonBodyPipe`use native json method

### Depricated
ServerRequest not more exists, use Request
Not support for deno 1.25.x and lower or not testet
Obsolate authentification example

## 0.2.12

Integrate plugin swagger and refactorings update swagger plugin with definitions
by prop

## 0.2.11

Bugfixes

## 0.2.10

Bugfixes and upgrade sources

## 0.2.9

Add more event types for route

- API_ADD_ROUTE
- AFTER_ROUTE_RESPONSE

## 0.2.8

Bugfixes

## 0.2.7

integrate new events

- BEFORE_ROUTE
- BEFORE_REQUEST
- ROUTE_NOT_FOUND

## 0.2.3 - 0.2.6

hotfixes and optimization

## v0.2.2

- integrate cache control attributes for file pipe as optional parameter
- optional parameter for filePipe for statuscode
- add some tests

## v0.2.1

- remove extend method for pattern map (`extendPatternMap`)
- integrate new `describe` attribute for KeyMatch params (example exists)
- add Hash pattern for KeyMap

## v0.2.0

- add file handling
- add functionality for breaking pipe process
- add new example to handle files
  [File example](https://deno.land/x/deno_api_server/example/static-file.ts)

## 0.0.18 > 0.1.0

- KeyMatch not support Number type anymore use EPatternTypes or the string value


example/main.ts
/**
 * Main example file to see how it works
 */
import {
  Api,
  EMethod,
  EPatternTypes,
  IServerConfig,
  KeyMatch,
  RequestError,
  Route,
  BreakPipe,
} from "../mod.ts";

import statusRoute from "../src/presets/routes/status.ts";
import healthzRoute from "../src/presets/routes/healthz.ts";
import jsonBodyPipe from "../src/presets/pipes/body/json-body.pipe.ts";
import redirectPipe from "../src/presets/pipes/process/redirect.pipe.ts";
import { EEvent } from "../src/definition/event.ts";
import RouteEvent from "../src/definition/events/route.event.ts";
import RequestEvent from "../src/definition/events/request.event.ts";
import filePipe from "../src/presets/pipes/process/file.pipe.ts";
import htmlPipe from "../src/presets/pipes/process/html.pipe.ts";

console.log(`Init deno api server`, Deno.version);

const serverConfig: IServerConfig = { port: 8080 };
const api = new Api(serverConfig);

/**
 * use global event to get information of regist routes
 */
addEventListener(EEvent.API_ADD_ROUTE, (event) => {
  if (event instanceof RouteEvent) {
    const { route } = <RouteEvent> event;
    console.log(`Add Route ${route.methods.join(",")} ${route.matcher.uri}`);
  }
});

// after route response
addEventListener(EEvent.AFTER_ROUTE_RESPONSE, event => {
  if (event instanceof RequestEvent) {
    const { response, request } = <RequestEvent> event;
    console.log(`Response ${response.status} for ${request.url} `);
  }
})

api
  // add preset route for status info
  .addRoute(statusRoute)
  // add preset route for healthcheck, similar to statusRoute but without body
  .addRoute(healthzRoute())
  .addRoute(
    new Route(EMethod.GET, "/public/break.html")
      .addPipe(htmlPipe("<html><body><b>test</b></body></html>")),
  )
  // new url pattern matcher
  .addRoute(
    new Route(
      EMethod.GET,
      new URLPattern({ pathname: '/pattern/:id/:name' })
    )
    .addPipe(ctx => {
      ctx.response.body = {
        msg: 'url pattern',
        id: ctx.match.params.id,
        name: ctx.match.params.name,
        //getId: ctx.match.params.get('id'),
      }
    })
  )
  // more complex file handling
  .addRoute(
    // complex exable for file pipeline
    new Route(
      EMethod.GET,
      new KeyMatch("/process/:file", { file: { type: "rest" } }),
    )
      // example how to break an pipe process
      .addPipe(async (ctx) => {
        const filePath = ctx.match.params.get("file") as string;
        return await filePipe(`/app/example/public/${filePath}`, {
          noThrow: true,
        })(ctx);
      })
      // will return fallback file of not found
      .addPipe(filePipe(`/app/example/not-found.jpg`, { statusCode: 404, contentType: "image/jpg" })),
  )
  // simple example to response hello for get request
  .addRoute(
    new Route(EMethod.GET, "/")
      .addPipe(({ response }) => {
        response.body = { message: "Hello API" };
      }),
  )
  // simple example for 200 on head requests
  .addRoute(
    new Route(EMethod.HEAD, "/"),
  )
  /**
   * use redirect preset to send redirection
   *
   * @preset redirectPipe
   */
  .addRoute(
    new Route(EMethod.GET, "/redirect")
      .addPipe(redirectPipe("/redirect-target?name=flex")),
  )
  // the target route on redirect
  .addRoute(
    new Route(EMethod.GET, "/redirect-target")
      .addPipe(({ response, url }) => {
        response.body = {
          message: "Hello API Redirect",
          name: url.searchParams.get("name") || "not-set",
        };
      }),
  )
  // simple example to mix method on one route
  .addRoute(
    new Route([EMethod.GET, EMethod.POST], "/mixed-hello")
      .addPipe(({ response, request }) => {
        response.body = { message: `Hello API by ${request.method}` };
      }),
  )
  /**
   * simple example integration for json body process
   *
   * @preset jsonBodyPipe
   */
  .addRoute(
    new Route([EMethod.POST], "/body/json")
      .addPipe(jsonBodyPipe)
      .addPipe(({ response, state }) => {
        response.body = {
          message: `Hello API json body`,
          body: state.get("body"), // get json body, created by jsonBodyPipe
          bodyType: state.get("bodyType"), // created on jsonBodyPipe
        };
      }),
  )
  // more complex example to use url params
  .addRoute(
    new Route(
      [EMethod.GET, EMethod.POST],
      // use the key matcher
      new KeyMatch(
        // define params
        "/get-by-key-name/:id/:name",
        // and describe uri params
        {
          id: { type: EPatternTypes.NUMBER }, // you can use 'Number' as string
          name: {},
        },
      ),
    )
      .addPipe(({ response, match }) => {
        const { params } = match;
        response.body = {
          message: `You call with keymatch`,
          id: params.get("id"),
          name: params.get("name"),
        };
      }),
  )
  // test example for long delay, all pipes support promisses
  .addRoute(
    new Route(EMethod.GET, "/wait")
      .addPipe(async ({ response, state }) => {
        state.set("begin", Date.now());
        return new Promise((resolve) => {
          setTimeout(resolve, 500);
        });
      })
      .addPipe(({ response, state }) => {
        response.body = {
          message: "Wait a while",
          begin: state.get("begin"),
          end: Date.now()
        };
      }),
  )
  /**
   * example to you multi pipes and pass data
   */
  .addRoute(
    new Route(EMethod.GET, "/state")
      .addPipe(({ response, state }) => {
        const before = new Date();
        response.body = { before };
        state.set("called", before);
      })
      .addPipe(({ response, state }) => {
        Object.assign(response.body, {
          passed: state.get("called"),
          done: new Date(),
        });
      }),
  )
  // simple example with error, always use request error to safe secure error handling
  .addRoute(
    new Route(EMethod.GET, "/error")
      .addPipe(() => {
        throw new RequestError("I dont like", 400);
      }),
  );

// start listen
console.log(`Start server localhost:${api.serverConfig.port}`);
await api.listen();
